programs/bench.c

```diff
@@ -90,440 +87,662 @@ static UTIL_time_t g_displayClock = UTIL_TIME_INITIALIZER;
-        U32 nbCompressionLoops = (U32)((2 MB) / (srcSize+1)) + 1;  /* initial conservative speed estimate */
-        U64 totalCTime=0, totalDTime=0;
-        U32 cCompleted=g_decodeOnly, dCompleted=0;
+    {   
+        U64 const crcOrig = (adv->mode == BMK_decodeOnly) ? 0 : XXH64(srcBuffer, srcSize, 0);
 #       define NB_MARKS 4
         const char* const marks[NB_MARKS] = { " |", " /", " =",  "\\" };
         U32 markNb = 0;
-
-        coolTime = UTIL_getTime();
         DISPLAYLEVEL(2, "\r%79s\r", "");
-        while (!cCompleted || !dCompleted) {
 
-            /* overheat protection */
-            if (UTIL_clockSpanMicro(coolTime) > ACTIVEPERIOD_MICROSEC) {
-                DISPLAYLEVEL(2, "\rcooling down ...    \r");
-                UTIL_sleep(COOLPERIOD_SEC);
-                coolTime = UTIL_getTime();
+        if (adv->mode != BMK_decodeOnly) {
+            BMK_initCCtxArgs cctxprep;
+            BMK_customReturn_t compressionResults;
+            int completed = 0;
+            U64 totalLoops = 0, totalTime = 0, fastest = (U64)(-1LL);
+            UTIL_time_t coolTime = UTIL_getTime();
+            cctxprep.ctx = ctx;
+            cctxprep.dictBuffer = dictBuffer;
+            cctxprep.dictBufferSize = dictBufferSize;
+            cctxprep.cLevel = cLevel;
+            cctxprep.comprParams = comprParams;
+            cctxprep.adv = adv;
+            /* Compression */
+            DISPLAYLEVEL(2, "%2s-%-17.17s :%10u ->\r", marks[markNb], displayName, (U32)srcSize);
+            if(adv->loopMode == BMK_timeMode) {
+                U64 maxTime = adv->nbSeconds * TIMELOOP_NANOSEC;
+                unsigned nbLoops = 1;
+                while(!completed) {
+                    /* Overheat protection */
+                    if (UTIL_clockSpanMicro(coolTime) > ACTIVEPERIOD_MICROSEC) {
+                        DEBUGOUTPUT("\rcooling down ...    \r");
+                        UTIL_sleep(COOLPERIOD_SEC);
+                        coolTime = UTIL_getTime();
+                    }
+
+                    compressionResults = BMK_benchFunction(&local_defaultCompress, (void*)ctx, &local_initCCtx, (void*)&cctxprep,
+                    nbBlocks, srcPtrs, srcSizes, cPtrs, cSizes, nbLoops);
+                    if(compressionResults.error) {
+                        results.error = compressionResults.error;
+                        return results;
+                    }
+
+                    {   U64 loopDuration = compressionResults.result.nanoSecPerRun * nbLoops;
+                        totalLoops += nbLoops;
+                        totalTime += loopDuration;
+                        if (loopDuration > 0) { // nanoSec / run
+                            fastest = MIN(fastest, compressionResults.result.nanoSecPerRun);
+                            nbLoops = (U32)(TIMELOOP_NANOSEC / fastest) + 1;
+                        } else {
+                            assert(nbLoops < 40000000);  /* avoid overflow */
+                            nbLoops *= 2;
```

Comment:
if `loopDuration == 0`, you'll want to increase `nbLoops` by more than a factor 2,
otherwise next measurement will likely be insignificant too.

Reply:
My thought here was that if loopDuration was 0, the execution time of the function is rather minimal, so the execution time of these loops is mostly dominated by the first time loopDuration isn't zero. For example, assuming a 10ms resolution, a 9 ms time could be potentially displayed as 0. With the previous 100x multiplier, this would mean the first nonzero loopDuration would be almost a second long, which means the total execution time could be close to 2 seconds even though we pass in nbSeconds = 1. The main downside is that the getTime functions would need to be called more times, but the overhead for that at most ~25 calls of that, which doesn't seem too large.